zson/lexer.go

```diff
@@ -300,6 +318,123 @@ func (l *Lexer) scanString() (string, error) {
 	}
 }
 
+// scanToCloseQuote scans the input into the slice appending to the b.
+// It peeks at but does not consume the final end quote character.
+func (l *Lexer) scanToCloseQuote(b []byte) ([]byte, error) {
+	for {
+		c, err := l.peek()
+		if err != nil {
+			return nil, err
+		}
+		if c == '"' {
+			return b, nil
+		}
+		l.skip(1)
+		if c == '\\' {
+			b = append(b, '\\')
+			c, err = l.readByte()
+			if err != nil {
+				return nil, err
+			}
+		}
+		b = append(b, c)
+	}
+}
+
+//XXX inspired by golang package json source
+func parseStringBytes(b *strings.Builder, bytes []byte) (string, error) {
+	k := 0
+	for k < len(bytes) {
+		switch c := bytes[k]; {
+		case c == '\\':
+			k++
+			if k >= len(bytes) {
+				panic("can't happen because string scanner would look for next char")
+			}
+			switch c := bytes[k]; c {
+			default:
+				return "", fmt.Errorf("illegal escape (\\%c) in string", c)
+			case '"', '\\', '/', '\'':
+				b.WriteByte(bytes[k])
+				k++
+			case 'b':
+				b.WriteByte('\b')
+				k++
+			case 'f':
+				b.WriteByte('\f')
+				k++
+			case 'n':
+				b.WriteByte('\n')
+				k++
+			case 'r':
+				b.WriteByte('\r')
+				k++
+			case 't':
+				b.WriteByte('\t')
+				k++
+			case 'u':
+				k++
+				r, err := unhexRune(bytes[k:])
+				if err != nil {
+					return "", err
+				}
+				k += 4
+				if utf16.IsSurrogate(r) {
+					if len(bytes) < 6 || bytes[0] != '\\' || bytes[1] != 'u' {
+						return "", errors.New("illegal surrage utf-16 rune pair")
+					}
+					r2, err := unhexRune(bytes[k+2:])
+					if err != nil {
+						return "", err
+					}
+					k += 6
+					if dec := utf16.DecodeRune(r, r2); dec != unicode.ReplacementChar {
+						// A valid pair; consume.
+						if _, err := b.WriteRune(dec); err != nil {
+							return "", err
+						}
+					}
+					//XXX
+					// Invalid surrogate; fall back to replacement rune.
+					//rr = unicode.ReplacementChar
```

Comment:
Should this just return an error?
```suggestion
					return "", errors.New("invalid utf16 surrogate pair")
```

Reply:
Yes, I made an issue.